### PR TITLE
Softmax layer

### PR DESCRIPTION
**Pull request by d-lowl**
Adds Softmax layer implementation and tests (#52).

Example repo (instructions included): https://github.com/d-lowl/fantastic-bassoon

Limitations:
* As discussed with zaleslaw, masks are not implemented, as they are out of scope for this issue. To be implemented later.
* ~~Only the last axis is supported for normilisation (i.e. the default axis). The reason -- Tensorflow Java lib's softmax implementation does not have axis parameter (as opposed to Python's implementation). If not needed at this point, I can remove the WIP and the PR can be merged. Otherwise, I'm working on it now.~~ See update

### CODE REVIEW

**Comment by d-lowl** [comment]
UPD: I've implemented handling normalisation along non-last axis, essentially back-porting this functionality from Tensorflow 2 (it's not available in 1.15 API). Should be good for review now.

For tests I've added another function for asserting on 2D inputs to activation layers. I was trying to generalise with the existing functions (by avan1235), but converting from Tensors/Operands to Array<*> conflicts with doing the same for FloatArrays. So I gave up on that.

**Comment by zaleslaw** [comment]
So, thanks for the PR and great refactoring of ModelSaver, please wait a few days with this PR before merging, there is a queue of PRs with changes in ModelSaver

Hope you will be closing this chain of PRs related to the activation layers and will make the same refactoring in the "when" op in ModelSaver? 

Also, I see a great example project with a demonstration of the Softmax layer, maybe later we will add it as an example

**Comment by d-lowl** [comment]
Yeah, cheers. 

> please wait a few days 
of course, no rush here

> and will make the same refactoring in the "when" op in ModelSaver?
If I remember correctly there are a couple more places where `when` clause can be switched from "class patterns" to "is pattern". But I haven't touched them, as unrelated to that particular layer. I guess they can also be switched.

**Comment by zaleslaw** [comment]
So, I agree with you, this kind of refactoring is out of scope but will be useful later

**Comment by zaleslaw** [comment]
So, @d-lowl it's your turn! Please merge the master, fix some refactoring issue and I'll run the TC after that

**Comment by d-lowl** [comment]
Merged, when clauses are refactored (in the last commit), good to go from my side 👌

**Comment by zaleslaw** [comment]
Great PR @d-lowl thanks for the contrib!
